File: core/src/com/jupiter/europa/screen/dialog/OptionsDialog.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.jupiter.europa.screen.dialog;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Slider;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.SpriteDrawable;
import com.jupiter.europa.EuropaGame;
import com.jupiter.europa.scene2d.ui.EuropaButton;
import com.jupiter.europa.scene2d.ui.EuropaButton.ClickEvent;
import com.jupiter.europa.scene2d.ui.ObservableDialog;
import com.jupiter.europa.scene2d.ui.TabbedPane;
import com.jupiter.europa.screen.MainMenuScreen;
import static com.jupiter.europa.screen.MainMenuScreen.DEFAULT_KEY;
import static com.jupiter.europa.screen.MainMenuScreen.TAB_STYLE_KEY;

/**
 *
 * @author Nathan Templon
 */
public class OptionsDialog extends ObservableDialog {

    // Enumerations
    public enum OptionsDialogExitStates {

        ACCEPT,
        CANCEL
    }


    // Constants
    private static final String DIALOG_NAME = "";


    // Static Methods
    private static Skin getSkin() {
        return MainMenuScreen.getMainMenuSkin();
    }


    // Fields
    private final Skin skin;

    private Table mainTable;
    private Label optionsLabel;
    private TabbedPane optionsPane;
    private EuropaButton optionsAcceptButton;
    private EuropaButton optionsCancelButton;
    private Table optionsButtonTable;

    private AudioOptionsTable audioTable;
    private GraphicOptionsTable graphicsTable;

    private OptionsDialogExitStates exitState;


    // Properties
    public final OptionsDialogExitStates getExitState() {
        return this.exitState;
    }


    // Initialization
    public OptionsDialog() {
        super(DIALOG_NAME, getSkin().get(WindowStyle.class));

        this.skin = getSkin();

        this.initComponent();

        this.addDialogListener((DialogEventArgs args) -> this.loadSettings(), DialogEvents.SHOWN);
    }


    // Private Methods
    private void initComponent() {
        this.mainTable = new Table();
        this.optionsLabel = new Label("Options", skin.get(DEFAULT_KEY, Label.LabelStyle.class));
        this.optionsPane = new TabbedPane(skin.get(TAB_STYLE_KEY, TextButton.TextButtonStyle.class));

        this.optionsAcceptButton = new EuropaButton("Accept", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
        this.optionsAcceptButton.addClickListener(this::onAcceptClick);

        this.optionsCancelButton = new EuropaButton("Cancel", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
        this.optionsCancelButton.addClickListener(this::onCancelClick);

        this.optionsButtonTable = new Table();
        this.optionsButtonTable.add(this.optionsCancelButton).space(MainMenuScreen.COMPONENT_SPACING).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).right()
                .expandX();
        this.optionsButtonTable.add(this.optionsAcceptButton).space(MainMenuScreen.COMPONENT_SPACING).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).right();

        // Create and Add Tabs
        this.audioTable = new AudioOptionsTable(this.skin);
        this.graphicsTable = new GraphicOptionsTable(this.skin);
        this.optionsPane.addTab("Audio", this.audioTable);
        this.optionsPane.addTab("Graphics", this.graphicsTable);

        this.mainTable.add(this.optionsLabel).left();
        this.mainTable.row();
        this.mainTable.add(this.optionsPane).expand().fill().top();
        this.mainTable.row();
        this.mainTable.add(this.optionsButtonTable).expandX().fillX().right();

        this.mainTable.pad(MainMenuScreen.TABLE_PADDING);
        this.mainTable.background(this.skin.get(MainMenuScreen.DIALOG_BACKGROUND_KEY, SpriteDrawable.class));

        this.getContentTable().add(this.mainTable).center().expandY().fillY().width(MainMenuScreen.DIALOG_WIDTH);
    }

    private void onAcceptClick(ClickEvent event) {
        this.applySettings();
        this.hide();
    }
    
    private void onCancelClick(ClickEvent event) {
        this.hide();
    }

    private void loadSettings() {
        this.audioTable.setMusicVolume(EuropaGame.game.getSettings().musicVolume.get());
    }

    private void applySettings() {
        EuropaGame.game.getSettings().musicVolume.set(this.audioTable.getMusicVolume());
        EuropaGame.game.saveSettings();
    }


    // Nested Classes
    private static class AudioOptionsTable extends Table {

        // Fields
        private final Skin skin;
        private final Label volumeLabel;
        private final Slider musicSlider;


        // Properties
        public final void setMusicVolume(float volume) {
            this.musicSlider.setValue(volume);
        }

        public final float getMusicVolume() {
            return this.musicSlider.getValue();
        }


        // Initialization
        private AudioOptionsTable(Skin skin) {
            this.skin = skin;

            this.volumeLabel = new Label("Volumne:", skin.get(MainMenuScreen.INFO_STYLE_KEY, LabelStyle.class));

            this.musicSlider = new Slider(0f, 1f, 0.05f, false, skin.get(DEFAULT_KEY, Slider.SliderStyle.class));

            this.add(this.volumeLabel).space(MainMenuScreen.COMPONENT_SPACING).center().left();
            this.add(this.musicSlider).space(MainMenuScreen.COMPONENT_SPACING).center().minHeight(20).expandX().fillX();
            this.row();

            this.pad(MainMenuScreen.TABLE_PADDING);
        }

    }


    private static class GraphicOptionsTable extends Table {


        // Fields
        private final Skin skin;


        // Initialization
        private GraphicOptionsTable(Skin skin) {
            this.skin = skin;

            this.pad(MainMenuScreen.TABLE_PADDING);
        }

    }

}

File: core/src/com/jupiter/europa/screen/dialog/OptionsDialog.kt
/*
 * The MIT License
 *
 * Copyright 2015 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package com.jupiter.europa.screen.dialog

import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.utils.SpriteDrawable
import com.jupiter.europa.EuropaGame
import com.jupiter.europa.scene2d.ui.EuropaButton
import com.jupiter.europa.scene2d.ui.EuropaButton.ClickEvent
import com.jupiter.europa.scene2d.ui.ObservableDialog
import com.jupiter.europa.scene2d.ui.TabbedPane
import com.jupiter.europa.screen.MainMenuScreen

import com.jupiter.ganymede.event.Listener

/**

 * @author Nathan Templon
 */
public class OptionsDialog : ObservableDialog(OptionsDialog.DIALOG_NAME, OptionsDialog.getDefaultSkin().get(javaClass<Window.WindowStyle>())) {

    // Enumerations
    public enum class OptionsDialogExitStates {

        ACCEPT,
        CANCEL
    }


    // Fields
    private val skinInternal: Skin

    private var mainTable: Table? = null
    private var optionsLabel: Label? = null
    private var optionsPane: TabbedPane? = null
    private var optionsAcceptButton: EuropaButton? = null
    private var optionsCancelButton: EuropaButton? = null
    private var optionsButtonTable: Table? = null

    private var audioTable: AudioOptionsTable? = null
    private var graphicsTable: GraphicOptionsTable? = null

    public val exitState: OptionsDialogExitStates? = null


    init {

        this.skinInternal = getDefaultSkin()

        this.initComponent()

        this.addDialogListener(Listener { this.loadSettings() }, ObservableDialog.DialogEvents.SHOWN)
    }


    // Private Methods
    private fun initComponent() {
        this.mainTable = Table()
        this.optionsLabel = Label("Options", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<Label.LabelStyle>()))
        this.optionsPane = TabbedPane(skinInternal.get(MainMenuScreen.TAB_STYLE_KEY, javaClass<TextButton.TextButtonStyle>()))

        this.optionsAcceptButton = EuropaButton("Accept", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
        this.optionsAcceptButton!!.addClickListener(Listener { args -> this.onAcceptClick(args) })

        this.optionsCancelButton = EuropaButton("Cancel", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
        this.optionsCancelButton!!.addClickListener(Listener { this.onCancelClick(it) })

        this.optionsButtonTable = Table()
        this.optionsButtonTable!!.add<EuropaButton>(this.optionsCancelButton).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).right().expandX()
        this.optionsButtonTable!!.add<EuropaButton>(this.optionsAcceptButton).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).right()

        // Create and Add Tabs
        this.audioTable = AudioOptionsTable(this.skinInternal)
        this.graphicsTable = GraphicOptionsTable(this.skinInternal)
        this.optionsPane!!.addTab("Audio", this.audioTable!!)
        this.optionsPane!!.addTab("Graphics", this.graphicsTable!!)

        this.mainTable!!.add<Label>(this.optionsLabel).left()
        this.mainTable!!.row()
        this.mainTable!!.add<TabbedPane>(this.optionsPane).expand().fill().top()
        this.mainTable!!.row()
        this.mainTable!!.add<Table>(this.optionsButtonTable).expandX().fillX().right()

        this.mainTable!!.pad(MainMenuScreen.TABLE_PADDING.toFloat())
        this.mainTable!!.background(this.skinInternal.get(MainMenuScreen.DIALOG_BACKGROUND_KEY, javaClass<SpriteDrawable>()))

        this.getContentTable().add<Table>(this.mainTable).center().expandY().fillY().width(MainMenuScreen.DIALOG_WIDTH.toFloat())
    }

    private fun onAcceptClick(event: ClickEvent) {
        this.applySettings()
        this.hide()
    }

    private fun onCancelClick(event: ClickEvent) {
        this.hide()
    }

    private fun loadSettings() {
        this.audioTable!!.setMusicVolume(EuropaGame.game.settings!!.musicVolume.get())
    }

    private fun applySettings() {
        EuropaGame.game.settings!!.musicVolume.set(this.audioTable!!.getMusicVolume())
        EuropaGame.game.saveSettings()
    }


    // Nested Classes
    private class AudioOptionsTable(private val skinInternal: Skin) : Table() {
        private val volumeLabel: Label
        private val musicSlider: Slider


        // Properties
        public fun setMusicVolume(volume: Float) {
            this.musicSlider.setValue(volume)
        }

        public fun getMusicVolume(): Float {
            return this.musicSlider.getValue()
        }


        init {

            this.volumeLabel = Label("Volumne:", skinInternal.get(MainMenuScreen.INFO_STYLE_KEY, javaClass<LabelStyle>()))

            this.musicSlider = Slider(0f, 1f, 0.05f, false, skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<Slider.SliderStyle>()))

            this.add(this.volumeLabel).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).center().left()
            this.add(this.musicSlider).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).center().minHeight(20f).expandX().fillX()
            this.row()

            this.pad(MainMenuScreen.TABLE_PADDING.toFloat())
        }

    }


    private class GraphicOptionsTable(private val skinInternal: Skin) : Table() {


        init {

            this.pad(MainMenuScreen.TABLE_PADDING.toFloat())
        }

    }

    companion object {


        // Constants
        private val DIALOG_NAME = ""


        // Static Methods
        private fun getDefaultSkin(): Skin {
            return MainMenuScreen.createMainMenuSkin()
        }
    }

}
